### PR TITLE
Esc 3

### PR DESCRIPTION
**Pull request by rvazarkar**
## Description

This MR implements the ADCSESC3 edge fully on the API side (post processing, edge composition). Also updates cross product logic to allow any number of input sets

## Motivation and Context

ESC3 is the second escalation covered by BloodHound, and this MR implements the pieces necessary to process it

## How Has This Been Tested?
New integration tests have been written to cover the changes. Additionally, testing was done manually using the BHE ADCS labs to ensure that expected results are produced

## Screenshots (if appropriate):
image
image

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: packages/go/analysis/ad/adcs.go [hunk 1]
```diff
@@ -377,55 +468,75 @@ func certTemplateHasEku(certTemplate *graph.Node, targetEkus ...string) (bool, e
...
 
-			return &operation.Stats, operation.Done()
+				return nil
+			})
+
+			operation.Operation.SubmitReader(func(ctx context.Context, tx graph.Transaction, outC chan<- analysis.CreatePostRelationshipJob) error {
+				for _, enterpriseCA := range enterpriseCertAuthorities {
+					if cache.DoesCAChainProperlyToDomain(enterpriseCA, innerDomain) {
+						if err := PostADCSESC3(ctx, tx, outC, groupExpansions, enterpriseCA, innerDomain, cache); err != nil {
+							log.Errorf("failed post processing for %s: %v", ad.ADCSESC1.String(), err)
```
Comment (juggernot325) [inline packages/go/analysis/ad/adcs.go:528]: Looks like a copy/paste error on the `ad.ADCSESC1` string.

**Comment by juggernot325** [comment]
Just a heads up that we'll need to update this line in the docs to indicate that the `ADCSESC3` edge can now be used with the `/api/v2/graphs/edge-composition` endpoint.

https://github.com/SpecterOps/BloodHound/blob/main/cmd/api/src/docs/json/paths/v2/graphs.json#L98

File: cmd/api/src/model/appcfg/flag.go [hunk 2]
```diff
@@ -76,10 +76,10 @@ func AvailableFlags() FeatureFlagSet {
 		},
 		FeatureAdcs: {
 			Key:           FeatureAdcs,
-			Name:          "Enable collection and processing of Active Directory Certificate Services Data",
-			Description:   "Enables the ability to collect, analyze, and explore Active Directory Certificate Services data and previews new attack paths.",
+			Name:          "Enable collection and processing of ADCS data",
+			Description:   "Enables the ability to analyze ADCS data for creating ADCS attack paths",
 			Enabled:       false,
-			UserUpdatable: true,
+			UserUpdatable: false,
```
Comment (juggernot325) [inline cmd/api/src/model/appcfg/flag.go:82]: Were these changes intended? Or are we removing the ability for users to update the feature flag with the release of ESC3?

Comment (rvazarkar) [reply cmd/api/src/model/appcfg/flag.go:82]: Nope, it was a bad merge

File: packages/go/analysis/ad/ad.go [hunk 3]
```diff
@@ -482,11 +539,268 @@ func GetEdgeCompositionPath(ctx context.Context, db graph.Database, edge *graph.
...
+				for _, p5 := range enterpriseCASegments[eca2.ID] {
+					paths.AddPath(p5.Path())
+				}
+
+				paths.AddPath(p3)
+				paths.AddPath(p1.Path())
+				paths.AddPath(p2.Path())
+
+				if collected, err := eca2.Properties.Get(ad.EnrollmentAgentRestrictionsCollected.String()).Bool(); err != nil {
+					log.Errorf("error getting enrollmentagentcollected for eca2 %d: %w", eca2.ID, err)
```
Comment (urangel) [inline packages/go/analysis/ad/ad.go:770]: The error wrapping here won't log well since we are trying to use the error directly so %w should be switched to %v when we are logging directly instead of returning an error up the call stack.

File: packages/go/analysis/ad/adcscache.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	log.Infof("Finished building adcs cache")
+}
+
+func nodeSetToBitmap(set graph.NodeSet) cardinality.Duplex[uint32] {
+	bitmap := cardinality.NewBitmap32()
+	for _, s := range set {
+		bitmap.Add(s.ID.Uint32())
+	}
+	return bitmap
+}
```
Comment (urangel) [inline packages/go/analysis/ad/adcscache.go:96]: Would `NodeSetToDuplex` work for getting a bitmap of ID's from a nodeset?

Comment (rvazarkar) [reply packages/go/analysis/ad/adcscache.go:96]: Sure would

File: packages/go/analysis/ad/ad.go (hunk 3 above)
Comment (rvazarkar) [reply packages/go/analysis/ad/ad.go:770]: Fixed

File: packages/go/analysis/ad/ad.go [hunk 5]
```diff
@@ -297,47 +297,55 @@ func createOrUpdateWellKnownLink(tx graph.Transaction, startNode *graph.Node, en
 	}
 }
 
-func CalculateCrossProductBitmaps(firstNodeSet, secondNodeSet cardinality.Duplex[uint32], groupExpansions impact.PathAggregator) cardinality.Duplex[uint32] {
+func CalculateCrossProductNodeSets(groupExpansions impact.PathAggregator, nodeSets ...[]*graph.Node) cardinality.Duplex[uint32] {
+	if len(nodeSets) < 2 {
+		log.Errorf("cross products require at least 2 nodesets")
```
[the remark this answers is not in the document]
Comment (rvazarkar) [reply packages/go/analysis/ad/ad.go:302]: Theoretically, we should never hit this case, so I'd like to leave it as an error

File: packages/go/analysis/ad/queries.go [hunk 6]
```diff
@@ -1479,3 +1491,13 @@ func FetchCertTemplatePathToDomain(tx graph.Transaction, certTemplate, domain *g
...
 	})
 }
+
+func FetchCertTemplateCAs(tx graph.Transaction, certTemplate *graph.Node) (graph.NodeSet, error) {
+	return ops.FetchEndNodes(tx.Relationships().Filter(
+		query.And(
+			query.Equals(query.StartID(), certTemplate.ID),
+			query.KindIn(query.Relationship(), ad.PublishedTo),
+			query.KindIn(query.End(), ad.EnterpriseCA),
+		),
```
[the remark this answers is not in the document]
Comment (rvazarkar) [reply packages/go/analysis/ad/queries.go:1501]: Fixed

File: packages/go/analysis/ad/adcscache.go (hunk 4 above)
Comment (urangel) [reply packages/go/analysis/ad/adcscache.go]: Nice 🙌

File: packages/go/analysis/ad/ad.go (hunk 3 above)
Comment (urangel) [reply packages/go/analysis/ad/ad.go]: Thank you! There are some more instances of this (I fixed a bunch in [this PR](https://github.com/SpecterOps/BloodHound/pull/292/files) that is being cherry picked into the upcoming patch). Would you mind combing through for any other instances where we are directly logging an error and using `%w` instead of `%v` please?

Comment (rvazarkar) [reply packages/go/analysis/ad/ad.go]: Just did a find/replace to hopefully get all of them

File: packages/go/analysis/ad/ad.go [hunk 7]
```diff
@@ -387,37 +395,77 @@ func CalculateCrossProductBitmaps(firstNodeSet, secondNodeSet cardinality.Duplex
 	return resultEntities
 }
 
-func CalculateCrossProductNodeSets(firstNodeSet, secondNodeSet []*graph.Node, groupExpansions impact.PathAggregator) cardinality.Duplex[uint32] {
+func CalculateCrossProductBitmaps(groupExpansions impact.PathAggregator, nodeSets ...cardinality.Duplex[uint32]) cardinality.Duplex[uint32] {
```
Comment (urangel) [inline packages/go/analysis/ad/ad.go:398]: Nice param reorder to allow for the variadic nodeSets param

File: packages/go/analysis/ad/ad.go [hunk 8]
```diff
@@ -387,37 +395,77 @@ func CalculateCrossProductBitmaps(firstNodeSet, secondNodeSet cardinality.Duplex
...
-		if entity.Kinds.ContainsOneOf(ad.Group) {
-			firstSetUnroll.Or(groupExpansions.Cardinality(entity.ID.Uint32()).(cardinality.Duplex[uint32]))
+	//Take the second of our node sets and unroll it all into a single bitmap
+	nodeSets[1].Each(func(id uint32) (bool, error) {
+		checkSet.Add(id)
+		idCardinality := groupExpansions.Cardinality(id)
+		idCardinalityCount := getCardinalityCount(id, idCardinality, cardinalityCache)
+		cardinalityCache[id] = idCardinalityCount
+		if idCardinalityCount > 0 {
+			cardinalityCache[id] = idCardinalityCount
```
Comment (urangel) [inline packages/go/analysis/ad/ad.go:419]: Line 417 seems like it might be redundant? `cardinalityCache[id] = idCardinalityCount`

**Review by urangel — APPROVED** [review]
Great work! Thanks for continuing to tend to the PR with all of the changes coming in

**Review by ddlees — APPROVED** [review]

**Review by benwaples — APPROVED** [review]
